src/workerd/api/sockets.c++

```diff
@@ -60,122 +69,23 @@ jsg::Ref<Socket> connectImpl(
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::abortRead() {
-  thenOrRunNow<void>([](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->abortRead();
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::getsockopt(int level, int option, void* value, uint* length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->getsockopt(level, option, value, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::setsockopt(int level, int option, const void* value, uint length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->setsockopt(level, option, value, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::getsockname(struct sockaddr* addr, uint* length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->getsockname(addr, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::getpeername(struct sockaddr* addr, uint* length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->getpeername(addr, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-kj::Promise<size_t> PipelinedAsyncIoStream::read(void* buffer, size_t minBytes, size_t maxBytes) {
-  return thenOrRunNow<size_t>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->read(buffer, minBytes, maxBytes);
-  });
-}
-
-kj::Promise<size_t> PipelinedAsyncIoStream::tryRead(void* buffer, size_t minBytes, size_t maxBytes) {
-  return thenOrRunNow<size_t>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->tryRead(buffer, minBytes, maxBytes);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::write(const void* buffer, size_t size) {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->write(buffer, size);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->write(pieces);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::whenWriteDisconnected() {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->whenWriteDisconnected();
-  });
+  // Forcibly close the readable/writable streams.
+  auto& context = IoContext::current();
+  context.addTask(context.run([this](jsg::Lock& js) {
```

Comment:
Why `context.addTask(context.run(`? This seems redundant, but forces exiting and re-entering the isolate. Can you just execute the code below inline?

Reply:
Here is why I ended up with this, it's possible there is another way to achieve each of these steps differently to avoid doing what I'm doing here:

* In order to call `cancel` and `abort` I need the `jsg::Lock`
* In order to get the `jsg::Lock` I need to call `context.run`
* The method doesn't return a Promise, so `context.addTask` on that promise so it doesn't get lost

As a side note, I think we will end up with `close` returning a Promise again. But we need to discuss this aspect of the API and make a decision so I'd rather leave it as-is in this PR.

Reply:
You can add a parameter `jsg::Lock& js` to `close()` itself, and JSG will automatically fill it in.

It might be a good idea to re-read the JSG docs.

Reply:
Ahh. I knew about this pattern but it didn't click in my head that `close` is a JSG_METHOD that can use it...